gobblin-modules/gobblin-gitter/src/main/java/org/apache/gobblin/source/GitterMessageExtractor.java

```diff
@@ -0,0 +1,270 @@
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gobblin.source;
+
+import java.io.IOException;
+import java.math.BigInteger;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Queue;
+
+import javax.annotation.Nullable;
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.avro.Schema;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.commons.io.IOUtils;
+import org.apache.gobblin.configuration.ConfigurationKeys;
+import org.apache.gobblin.configuration.WorkUnitState;
+import org.apache.gobblin.converter.DataConversionException;
+import org.apache.gobblin.converter.avro.JsonRecordAvroSchemaToAvroConverter;
+import org.apache.gobblin.password.PasswordManager;
+import org.apache.gobblin.source.extractor.DataRecordException;
+import org.apache.gobblin.source.extractor.Extractor;
+import org.apache.gobblin.source.extractor.extract.BigIntegerWatermark;
+import org.apache.gobblin.source.workunit.WorkUnit;
+import org.apache.http.HttpResponse;
+import org.apache.http.ParseException;
+import org.apache.http.client.HttpClient;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.util.EntityUtils;
+
+import com.google.common.io.Closer;
+import com.google.gson.Gson;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonParser;
+
+
+/**
+ * Extractor to extract Gitter Messages.
+ */
+@Slf4j
+public class GitterMessageExtractor implements Extractor<Schema, GenericRecord> {
+
+  protected static final String GITTER_ROOM_NAMES_KEY = "source.gitter.roomNames";
+  protected static final String GITTER_EXTRACTOR_KEY = "source.gitter.extractorClass";
+  protected static final String GITTER_EXTRACTOR_DEFAULT = "org.apache.gobblin.source.GitterMessageExtractor";
+  protected static final String GITTER_REST_API_ENDPOINT_KEY = "source.gitter.restEndPoint";
+  protected static final String GITTER_REST_API_ENDPOINT_DEFAULT = "https://api.gitter.im/v1/rooms";
+
+  protected static final Gson gson = new Gson();
+  protected final Closer closer = Closer.create();
+  protected final WorkUnitState workUnitState;
+  protected final WorkUnit workUnit;
+  protected final String gitterKey;
+  protected Schema gitterMessageSchema;
+  protected BigInteger gitterRollingWatermark;
+
+  protected final Queue<GenericRecord> gitterMessages;
+
+  protected final String gitterRoom;
+  protected String gitterRoomId;
+  protected final String gitterRestEndpoint;
+  protected final HttpClient httpClient;
```

Comment:
Shall we guarantee this is a ClosableHttpClient because you have added it into closer?